From 0d9d3f7a35516cfa9f1a224f1ecad0ac963e4cb3 Mon Sep 17 00:00:00 2001
From: cryptite <cryptite@gmail.com>
Date: Tue, 17 Dec 2019 15:29:44 -0600
Subject: [PATCH] Dont recalculate permissions on world change

Often, but not 100% of the time, the server will reconstruct and recalculate the Command Map sent to players
when they change worlds. This is often laggy; worse-so depending on the extent of the tab-completed commands to be sent.
In our case, the map sent to the player on login is sufficient and is not tossed by the client for world changes.

As a result, we opt not to recalculate this on world change and this prevents a fairly heavy, and consistent main-thread
lagspike whenever players change worlds, which can be a lot.

diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index 1d7d1ffbf..df1003688 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -425,4 +425,9 @@ public class PaperConfig {
             */
         }
     }
+
+    public static boolean recalculatePermsOnWorldChange = true;
+    private static void recalculatePermsOnWorldChange() {
+        recalculatePermsOnWorldChange = getBoolean("settings.recalculate-perms-on-world-change", recalculatePermsOnWorldChange);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 7b79ee4fe..f162c963a 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -712,7 +712,9 @@ public abstract class PlayerList {
         entityplayer1.playerConnection.sendPacket(new PacketPlayOutServerDifficulty(worlddata.getDifficulty(), worlddata.isDifficultyLocked()));
         entityplayer1.playerConnection.sendPacket(new PacketPlayOutExperience(entityplayer1.exp, entityplayer1.expTotal, entityplayer1.expLevel));
         this.a(entityplayer1, worldserver);
+        movedToWorld = true; // Paper
         this.d(entityplayer1);
+        movedToWorld = false; // Paper
         if (!entityplayer.playerConnection.isDisconnected()) {
             worldserver.addPlayerRespawn(entityplayer1);
             this.players.add(entityplayer1);
@@ -754,6 +756,8 @@ public abstract class PlayerList {
         // CraftBukkit end
         return entityplayer1;
     }
+    
+    private boolean movedToWorld = false; // Paper
 
     public void d(EntityPlayer entityplayer) {
         GameProfile gameprofile = entityplayer.getProfile();
@@ -907,7 +911,7 @@ public abstract class PlayerList {
             entityplayer.playerConnection.sendPacket(new PacketPlayOutEntityStatus(entityplayer, b0));
         }
 
-        entityplayer.getBukkitEntity().recalculatePermissions(); // CraftBukkit
+        if(!(movedToWorld && !com.destroystokyo.paper.PaperConfig.recalculatePermsOnWorldChange)) entityplayer.getBukkitEntity().recalculatePermissions(); // CraftBukkit // Paper
         this.server.getCommandDispatcher().a(entityplayer);
     }
 
-- 
2.21.1 (Apple Git-122.3)

